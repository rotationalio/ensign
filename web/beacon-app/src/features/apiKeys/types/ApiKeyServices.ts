export interface APIKey {
	id: string;
    client_id: string;
    client_secret: string;
    name: string;
    owner: string;
    permissions: string[];
    created: string;
    modified: string;
}

export type NewAPIKey = Omit<APIKey, 'id'>;