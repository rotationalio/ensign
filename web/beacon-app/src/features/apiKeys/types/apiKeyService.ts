export interface APIKey {
  id: string;
  client_id: string;
  client_secret: string;
  name: string;
  owner: string;
  permissions: string[];
  created?: string;
  modified?: string;
}

export interface APIKeysQuery {
  getApiKeys: () => void;
  apiKeys: any;
  hasApiKeysFailed: boolean;
  wasApiKeysFetched: boolean;
  isFetchingApiKeys: boolean;
  error: any;
}

export type NewAPIKey = Omit<APIKey, 'id'>;
