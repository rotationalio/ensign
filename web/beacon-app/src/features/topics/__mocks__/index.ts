export const getTopicsStatsMockData = () => {
  const data = [
    {
      name: 'publishers',
      value: 2,
    },
    {
      name: 'subscribers',
      value: 3,
    },
    {
      name: 'total_events',
      value: 1000000,
    },
    {
      name: 'storage',
      value: 203,
      units: 'MB',
    },
  ];
  return data;
};

export const getTopicEventsMockData = () => {
  return [
    {
      type: 'Document',
      version: '1.0.0',
      mimetype: 'application/json',
      events: {
        value: 12345678,
        percent: 96.0,
      },
      storage: {
        value: 512,
        units: 'MB',
        percent: 98.5,
      },
    },
    {
      type: 'Feed Item',
      version: '0.8.1',
      mimetype: 'application/rss',
      events: {
        value: 98765,
        percent: 4.0,
      },
      storage: {
        value: 4.3,
        units: 'KB',
        percent: 1.5,
      },
    },
  ];
};

export const createBinaryFixture = () => {
  const binaryData = new Uint8Array([72, 101, 108, 108, 111]);

  // Return the binary data as ArrayBuffer
  return binaryData.buffer;
};

export const getXMLFixture = () => {
  const xmlData = `
  <distributedSystem>
  <node id="node-1">
    <name>Node 1</name>
    <type>Application Server</type>
    <ipAddress>192.168.0.101</ipAddress>
    <port>8080</port>
    <status>Online</status>
    <connectedNodes>
      <nodeRef>node-2</nodeRef>
      <nodeRef>node-3</nodeRef>
    </connectedNodes>
  </node>

  <node id="node-2">
    <name>Node 2</name>
    <type>Database Server</type>
    <ipAddress>192.168.0.102</ipAddress>
    <port>3306</port>
    <status>Online</status>
    <connectedNodes>
      <nodeRef>node-1</nodeRef>
    </connectedNodes>
  </node>

  <node id="node-3">
    <name>Node 3</name>
    <type>Load Balancer</type>
    <ipAddress>192.168.0.103</ipAddress>
    <port>80</port>
    <status>Online</status>
    <connectedNodes>
      <nodeRef>node-1</nodeRef>
      <nodeRef>node-4</nodeRef>
      <nodeRef>node-5</nodeRef>
    </connectedNodes>
  </node>

  <node id="node-4">
    <name>Node 4</name>
    <type>Microservice</type>
    <ipAddress>192.168.0.104</ipAddress>
    <port>5000</port>
    <status>Offline</status>
    <connectedNodes>
      <nodeRef>node-3</nodeRef>
    </connectedNodes>
  </node>

  <node id="node-5">
    <name>Node 5</name>
    <type>Microservice</type>
    <ipAddress>192.168.0.105</ipAddress>
    <port>5000</port>
    <status>Online</status>
    <connectedNodes>
      <nodeRef>node-3</nodeRef>
    </connectedNodes>
  </node>
</distributedSystem>
`;
  return xmlData;
};
