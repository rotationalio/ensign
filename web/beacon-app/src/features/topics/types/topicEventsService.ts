export interface TopicEvents {
  type: string;
  version: string;
  mimetype: string;
  events: {
    value: number;
    percent: number;
  };
  storage: {
    value: number;
    units: string;
    percent: number;
  };
}

export interface TopicEventsQuery {
  getTopicEvents: () => void;
  hasTopicEventsFailed: boolean;
  isFetchingTopicEvents: boolean;
  topicEvents: TopicEvents[];
  wasTopicEventsFetched: boolean;
  error: any;
}
