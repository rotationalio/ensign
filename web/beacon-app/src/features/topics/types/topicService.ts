import { QuickViewData } from '@/hooks/useFetchQuickView/quickViewService';
export interface Topic {
  id: string;
  topic_name: string;
  status: string;
  created?: string;
  modified?: string;
  publishers?: number;
  subscribers?: number;
  data_storage?: QuickViewData;
}

export interface TopicsResponse {
  project_id: string;
  topics: Topic[];
  prev_page_token: string;
  next_page_token: string;
}

export interface TopicsQuery {
  getTopics: () => void;
  topics: TopicsResponse;
  hasTopicsFailed: boolean;
  wasTopicsFetched: boolean;
  isFetchingTopics: boolean;
  error: any;
}

export interface TopicQuery {
  getTopic: () => void;
  topic: Topic;
  hasTopicFailed: boolean;
  wasTopicFetched: boolean;
  isFetchingTopic: boolean;
  error: any;
}

export interface TopicQuickViewResponse {
  data: IStats[];
}

export interface TopicStatsQuery {
  getTopicStats: () => void;
  hasTopicStatsFailed: boolean;
  isFetchingTopicStats: boolean;
  topicStats: IStats[];
  wasTopicStatsFetched: boolean;
  error: any;
}
