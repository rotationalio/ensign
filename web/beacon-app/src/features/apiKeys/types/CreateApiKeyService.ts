import { APIKey, NewAPIKey } from './ApiKeyServices';

export interface APIKeyMutation {
  createNewKey(key: NewAPIKey): void;
  reset(): void;
  key: APIKey;
  hasKeyFailed: boolean;
  wasKeyCreated: boolean;
  isCreatingKey: boolean;
}

export const isKeyCreated = (mutation: APIKeyMutation): mutation is Required<APIKeyMutation> =>
  mutation.wasKeyCreated && mutation.key != undefined;
