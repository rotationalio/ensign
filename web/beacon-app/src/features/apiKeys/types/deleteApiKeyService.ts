import { UseMutateFunction } from '@tanstack/react-query';

import { APIKey } from './apiKeyService';

export interface DeleteAPIKeyMutation {
  deleteApiKey: UseMutateFunction<unknown, unknown, void, unknown>;
  reset(): void;
  key: APIKey;
  hasKeyDeletedFailed: boolean;
  wasKeyDeleted: boolean;
  isDeletingKey: boolean;
  error: any;
}

export type APIKeyDTO = {
  topicID: string;
};

export const isKeyDeleted = (
  mutation: DeleteAPIKeyMutation
): mutation is Required<DeleteAPIKeyMutation> =>
  mutation.wasKeyDeleted && mutation.key != undefined;
