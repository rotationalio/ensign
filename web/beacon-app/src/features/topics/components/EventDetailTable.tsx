import { Trans } from '@lingui/macro';
import { Table } from '@rotational/beacon-core';
import { ErrorBoundary } from '@sentry/react';
import { useMemo } from 'react';
import { useParams } from 'react-router-dom';

import { getTopicEventsMockData } from '../__mocks__';
import { useFetchTopicEvents } from '../hooks/useFetchTopicEvents';
import { getEventDetailColumns } from '../utils';

const EventDetailTable = () => {
  const param = useParams();
  const { id: topicID } = param as { id: string };
  const { topicEvents, isFetchingTopicEvents } = useFetchTopicEvents(topicID);
  const initialColumns = useMemo(() => getEventDetailColumns(), []) as any;

  return (
    <div className="mx-4">
      <ErrorBoundary
        fallback={
          <div className="item-center my-auto flex w-full text-center font-bold text-danger-500">
            <p>
              <Trans>
                Sorry we are having trouble listing event details for your topic, please refresh the
                page and try again.
              </Trans>
            </p>
          </div>
        }
      >
        <Table
          columns={initialColumns}
          data-testId="event-detail-table"
          data={topicEvents || getTopicEventsMockData()} // for now we are using mock data until we have the API ready
          isLoading={isFetchingTopicEvents}
          data-cy="event-detail-table"
        />
      </ErrorBoundary>
    </div>
  );
};

export default EventDetailTable;
