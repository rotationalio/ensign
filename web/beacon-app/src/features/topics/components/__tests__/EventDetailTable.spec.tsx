import { render, screen } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';

import { getTopicEventsMockData } from '../../__mocks__';
import { TopicEvents } from '../../types/topicEventsService';
import EventDetailTable from '../EventDetailTable';
// Mock the custom hook `useFetchTopicEvents`
vi.mock('../../hooks/useFetchTopicEvents', () => ({
  useFetchTopicEvents: () => ({
    topicEvents: getTopicEventsMockData(),
    isFetchingTopicEvents: false,
  }),
}));

const renderComponent = () => {
  return render(<EventDetailTable />);
};

// mock getEventDetailColumns
vi.mock('../../utils', async (importOrginial: any) => ({
  ...importOrginial,
  getEventDetailColumns: () => [
    {
      Header: `Event Type`,
      accessor: 'type',
    },
    {
      Header: `Version`,
      accessor: 'version',
    },
    {
      Header: `MIME Type`,
      accessor: 'mimetype',
    },
    // all columns with accessor function are not working yet
    // TODO: fix this later
    {
      Header: `# of Events`,
      accessor: (event: TopicEvents) => {
        return `${event?.events?.value}`;
      },
    },
    {
      Header: `% of Events`,
      accessor: (event: TopicEvents) => {
        return `${event?.events?.percent}%`;
      },
    },
    {
      Header: `Storage Volume`,
      accessor: (event: TopicEvents) => {
        return `${event?.storage?.value} ${event?.storage?.units}`;
      },
    },
    {
      Header: `% of Volume`,
      accessor: (event: TopicEvents) => {
        return `${event?.storage?.percent}%`;
      },
    },
  ],
}));

const MockTable = ({ columns, data, isLoading }: any) => {
  return (
    <table data-testid="event-detail-table">
      <thead>
        <tr>
          {columns.map((column) => (
            <th key={column.Header}>{column.Header}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {isLoading && <div>Loading...</div>}
        {data.map((row, rowIndex) => (
          <tr key={rowIndex}>
            {columns.map((column: any) => (
              <td key={column.accessor}>{row[column.accessor]}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
};
vi.mock('react-router-dom', async (importOrginial: any) => ({
  ...importOrginial,
  useNavigate: () => vi.fn(),
  useParams: () => ({ id: '1' }),
}));
// mock
//  mock table from beacon-core
vi.mock('@rotational/beacon-core', async (importOrginial: any) => ({
  ...importOrginial,
  Table: (props) => <MockTable {...props} />,
}));

vi.mock('@lingui/macro', async (importOrginial: any) => ({
  ...importOrginial,
  t: () => vi.fn(),
  Trans: () => vi.fn(),
}));

describe('EventDetailTable', () => {
  it('should render the component with default value', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render the correct columns', () => {
    renderComponent();
    expect(screen.getByText('Event Type')).toBeInTheDocument();
    expect(screen.getByText('Version')).toBeInTheDocument();
    expect(screen.getByText('MIME Type')).toBeInTheDocument();
    expect(screen.getByText('# of Events')).toBeInTheDocument();
    expect(screen.getByText('% of Events')).toBeInTheDocument();
    expect(screen.getByText('Storage Volume')).toBeInTheDocument();
    expect(screen.getByText('% of Volume')).toBeInTheDocument();
  });

  it('should contain the correct value', () => {
    renderComponent();
    expect(screen.getByText('Document')).toBeInTheDocument();
    expect(screen.getByText('application/json')).toBeInTheDocument();
  });
});
