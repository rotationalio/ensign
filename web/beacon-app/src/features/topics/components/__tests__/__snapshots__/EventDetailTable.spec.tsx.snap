// Vitest Snapshot v1

exports[`EventDetailTable > should render the component with default value 1`] = `
<div>
  <div
    class="mx-4"
  >
    <table
      data-testid="event-detail-table"
    >
      <thead>
        <tr>
          <th>
            Event Type
          </th>
          <th>
            Version
          </th>
          <th>
            MIME Type
          </th>
          <th>
            # of Events
          </th>
          <th>
            % of Events
          </th>
          <th>
            Storage Volume
          </th>
          <th>
            % of Volume
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Document
          </td>
          <td>
            1.0.0
          </td>
          <td>
            application/json
          </td>
          <td />
          <td />
          <td />
          <td />
        </tr>
        <tr>
          <td>
            Feed Item
          </td>
          <td>
            0.8.1
          </td>
          <td>
            application/rss
          </td>
          <td />
          <td />
          <td />
          <td />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
